File: Web/app/js/services/pollService.js
'use strict';

angular.module('yieldtome.services')

.service('PollService', ['$q', '$log', '$http', 'ConfigService',
    function($q, $log, $http, ConfigService) {

        this.deletePoll = function(poll) {
            $log.debug('Attempting to delete Poll');
            var deferred = $q.defer();

            if (poll == null) // Poll is not provided
            {
                var error = 'A Poll is required';
                $log.warn(error);
                deferred.reject(error);
                return deferred.promise;
            }

            // DELETE Polls/{pollID}
            var url = ConfigService.apiUrl + 'Polls/' + poll.PollID;
            $log.debug('Request Url: ' + url);

            $http.delete(url).success(function(data) {
                $log.debug('Poll deleted successfully.');
                deferred.resolve(data);
            })
            .error(function(status) { // Otherwise, some unknown error occured
                var error = status.Message;
                $log.warn(error);
                deferred.reject(error);
            });

            return deferred.promise;
        };

        this.updatePoll = function(poll) {
            $log.debug('Attempting to update Poll');
            var deferred = $q.defer();

            if (poll == null) {
                var error = 'A Poll object is needed';
                $log.warn(error);
                deferred.reject(error);
                return deferred.promise;
            } else if (poll.Name == null || poll.CreatorID == null) {
                var error = 'Both Name and CreatorID are mandatory fields on a Speakers List';
                $log.warn(error);
                deferred.reject(error);
                return deferred.promise;
            }

            $log.debug('Attempting to update the Poll: ' + poll.Name);

            var url = ConfigService.apiUrl + 'Polls/';

            $log.debug('Request Url: ' + url);

            $http.put(url, poll).success(function(data) {
                $log.debug('Updated Poll with PollID: ' + data.pollID);
                deferred.resolve(data);
            }).error(function(status) { // Otherwise, some unknown error occured
                var error = status.Message;
                $log.warn(error);
                deferred.reject(error);
            });

            return deferred.promise;
        };

        this.getPoll = function(pollID) {

            $log.debug('Attempting to get Poll associated with PollID: ' + pollID);
            var deferred = $q.defer();

            if (pollID == null || pollID == 0) {
                var error = 'PollID must be provided';
                $log.warn(error);
                deferred.reject(error);
                return deferred.promise;
            }

            var url = ConfigService.apiUrl + 'Polls/' + pollID;
            $log.debug('Request Url: ' + url);

            $http.get(url).success(function(data) {

                if (data == "null") // No List yet exists associated with this SpeakersListID
                {
                    var error = 'No Poll associated with PollID: ' + pollID;
                    $log.debug(error);
                    deferred.resolve(null);
                } else // Otherwise, there is a SpeakersList associated with this SpeakersListID
                {
                    deferred.resolve(data);
                }

            }).error(function(status) { // Otherwise, some unknown error occured
                var error = status.Message;
                $log.warn(error);
                deferred.reject(error);
            });

            return deferred.promise;
        };

        this.createPoll = function(event, poll, doAbstainsCount) {
            $log.debug('Attempting to create a new Speakers List');
            var deferred = $q.defer();

            if (event == null || event.EventID == null) {
                var error = 'An Event object with an EventID is needed';
                $log.warn(error);
                deferred.reject(error);
                return deferred.promise;
            } else if (poll == null) {
                var error = 'A Poll object is needed';
                $log.warn(error);
                deferred.reject(error);
                return deferred.promise;
            } else if (poll.Name == null || poll.CreatorID == null) {
                var error = 'Both Name and CreatorID are mandatory fields on a Poll';
                $log.warn(error);
                deferred.reject(error);
                return deferred.promise;
            }

            $log.debug('Attempting to create the Poll: ' + poll.Name);

            // POST Polls?eventID={eventID}&name={name}&creatorID={creatorID}&majorityRequired={majorityRequired}&doAbstainsCount={doAbstainsCount}
            var url = ConfigService.apiUrl + 'Polls?eventID=' + event.EventID +
                                            '&name=' + poll.Name +
                                            '&creatorID=' + poll.CreatorID;

            if(poll.MajorityRequired != null) { url = url + '&majorityRequired=' + poll.MajorityRequired; }
            if(doAbstainsCount) { url = url + '&doAbstainsCount=' + doAbstainsCount; }

            $log.debug('Request Url: ' + url);

            $http.post(url).success(function(data) {
                $log.debug('New Poll created with PollID: ' + data.PollID);
                deferred.resolve(data);
            }).error(function(status) { // Otherwise, some unknown error occured
                var error = status.Message;
                $log.warn(error);
                deferred.reject(error);
            });

            return deferred.promise;
        };

        this.getPolls = function(event) {
            $log.debug('Attempting to get Polls');
            var deferred = $q.defer();

            if (event == null) // Event is not provided
            {
                var error = 'An Event is required';
                $log.warn(error);
                deferred.reject(error);
                return deferred.promise;
            }

            var url = ConfigService.apiUrl + 'Events/' + event.EventID + '/Polls';
            $log.debug('Request Url: ' + url);

            $http.get(url).success(function(data) {
                $log.debug('Successfully retrieved ' + data.length + ' Polls');
                deferred.resolve(data);
            })
                .error(function(status) { // Otherwise, some unknown error occured
                    var error = status.Message;
                    $log.warn(error);
                    deferred.reject(error);
                });

            return deferred.promise;
        };

    }
]);

File: API/API.Data/Services/PollService.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.Composition;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using yieldtome.API.Data.Objects;
using yieldtome.Enums;
using yieldtome.Interfaces;

namespace yieldtome.API.Data.Services
{
    [Export(typeof(IPollService))]
    public class PollService : IPollService
    {
        [Import]
        IVoteService _voteService;

        private yieldtome.Objects.Poll CreatePollObject(Poll dbPoll)
        {
            yieldtome.Objects.Poll poll = new yieldtome.Objects.Poll()
            {
                PollID = dbPoll.PollID,
                Name = dbPoll.Name,
                MajorityRequired = dbPoll.MajorityRequired,
                Type = GetType(dbPoll.Type),
                CreatorID = dbPoll.CreatorID,
                Status = GetStatus(dbPoll.Status)
            };

            // Calculate Vote Counts
            List<yieldtome.Objects.Vote> votes = _voteService.GetVotes(poll.PollID);
            poll.VotesFor = votes.Where(x => x.VoteResult == Enums.VoteType.For).Count();
            poll.VotesAgainst = votes.Where(x => x.VoteResult == Enums.VoteType.Against).Count();
            poll.VotesAbstaining = votes.Where(x => x.VoteResult == Enums.VoteType.Abstain).Count();
            poll.IsPassing = IsPollPassing(poll);

            return poll;
        }

        private bool IsPollPassing(yieldtome.Objects.Poll poll)
        {
            bool isPassing = false;

            if (poll.VotesFor == 0) // No one has voted for
                isPassing = false;
            else if (poll.Type == PollType.ForMoreThanAgainst)
                if ((poll.VotesFor / (double)(poll.VotesFor + poll.VotesAgainst)) > (poll.MajorityRequired / 100.00))
                    isPassing = true;
                else // Number of for votes less than majority required
                    isPassing = false;
            else // PollType is ForMoreThanAgainstAndAbstain
                // Tests to see if number of For votes are greater than majority of total votes cast
                if ((poll.VotesFor / (double)(poll.VotesAbstaining + poll.VotesAgainst + poll.VotesFor) > 
                    (poll.MajorityRequired / 100.00)))
                    isPassing = true;

            return isPassing;
        }

        private SpeakersListAndPollStatus GetStatus(string statusString)
        {
            if (statusString.ToLower() == "closed")
                return SpeakersListAndPollStatus.Closed;
            else return SpeakersListAndPollStatus.Open;
        }

        private PollType GetType(string typeString)
        {
            if (typeString.ToLower() == "formorethanagainst")
                return PollType.ForMoreThanAgainst;
            else return PollType.ForMoreThanAgainstAndAbstain;
        }


        public List<yieldtome.Objects.Poll> GetPolls(int eventID)
        {
            Logging.LogWriter.Write(String.Format("Attempting to retrieve Polls for Event with EventID={0}", eventID));

            List<yieldtome.Objects.Poll> polls;
            using (var db = new Database())
            {
                Event dbEvent = db.Events.FirstOrDefault(x => x.EventID == eventID);

                if (dbEvent == null)
                {
                    string message = String.Format("No Event with EventID={0} exists", eventID);
                    throw new ArgumentException(message, "eventID");
                }

                polls = dbEvent.Polls
                                .Where(x => x.DeletedTime == null)
                                .Select(x => CreatePollObject(x)).ToList();
            }

            Logging.LogWriter.Write(String.Format("Successfully retrieved {0} Polls attending Event with EventID={1}", polls.Count, eventID));
            return polls;
        }

        public yieldtome.Objects.Poll GetPoll(int pollID)
        {
            Logging.LogWriter.Write(String.Format("Attempting to retrieve Poll with PollID={0}", pollID));

            yieldtome.Objects.Poll list;
            using (var db = new Database())
            {
                List<Poll> dbLists = db.Polls.Where(x => x.DeletedTime == null).ToList();
                Poll dbList = dbLists.FirstOrDefault(x => x.PollID == pollID);

                if (dbList == null)
                {
                    Logging.LogWriter.Write(String.Format("No Poll with PollID={0} exists", pollID));
                    return null;
                }

                list = CreatePollObject(dbList);
            }

            Logging.LogWriter.Write(String.Format("Successfully retrieved Poll with PollID={0}", pollID));
            return list;
        }

        public yieldtome.Objects.Poll CreatePoll(int eventID, string name, int creatorID, int majorityRequired, bool doAbstainsCount)
        {
            Logging.LogWriter.Write("Attempting to create a new Poll");

            if (name == "") throw new ArgumentNullException("Name is required", "name");
            if (majorityRequired > 100 || majorityRequired < 0) throw new ArgumentException("MajorityRequired must be between 0 and 100", "majorityRequired");

            if (GetPolls(eventID).Select(x => x.Name).Contains(name))
                throw new ArgumentException(String.Format("A Poll named {0} already exists", name), "name");

            Poll dbPoll;
            using (var db = new Database())
            {
                Profile profile = db.Profiles.FirstOrDefault(x => x.ProfileID == creatorID);
                if (profile == null) throw new ArgumentException(String.Format("No Profile with ProfileID={0} exists", creatorID));

                Event theEvent = db.Events.FirstOrDefault(x => x.EventID == eventID);
                if (theEvent == null) throw new ArgumentException(String.Format("No Event with EventID={0} exists", eventID));
                dbPoll = new Poll()
                {
                    Name = name,
                    Status = "Open",
                    MajorityRequired = majorityRequired,
                    EventID = eventID,
                    CreatorID = creatorID,
                    CreatedTime = DateTime.Now
                };

                if (doAbstainsCount) dbPoll.Type = PollType.ForMoreThanAgainstAndAbstain.ToString();
                else dbPoll.Type = PollType.ForMoreThanAgainst.ToString();

                db.Polls.Add(dbPoll);
                db.SaveChanges();
            }

            Logging.LogWriter.Write(String.Format("Created a new Poll with ID {0}", dbPoll.PollID));
            return CreatePollObject(dbPoll);
        }

        public yieldtome.Objects.Poll UpdatePoll(yieldtome.Objects.Poll updatedPoll)
        {
            Logging.LogWriter.Write(String.Format("Attempting to update Poll with PollID={0}", updatedPoll.PollID));

            if (updatedPoll.Name == "") throw new ArgumentNullException("updatedPoll.Name", "Name is required");
            if (updatedPoll.MajorityRequired > 100 || updatedPoll.MajorityRequired < 0) throw new ArgumentException("MajorityRequired must be between 0 and 100", "majorityRequired");

            using (var db = new Database())
            {
                Poll dbPoll = db.Polls.FirstOrDefault(x => x.PollID == updatedPoll.PollID);
                if (dbPoll == null) throw new ArgumentException(String.Format("No Poll with PollID={0} exists", updatedPoll.PollID));

                List<yieldtome.Objects.Poll> otherLists = GetPolls(dbPoll.EventID).Where(x => x.PollID != dbPoll.PollID).ToList();
                if (otherLists.Select(x => x.Name).Contains(updatedPoll.Name))
                    throw new ArgumentException(String.Format("A Poll named {0} already exists", updatedPoll.Name), "name");

                dbPoll.Name = updatedPoll.Name;
                dbPoll.Status = updatedPoll.Status.ToString();
                dbPoll.MajorityRequired = updatedPoll.MajorityRequired;
                dbPoll.Type = updatedPoll.Type.ToString();
                dbPoll.UpdatedTime = DateTime.Now;

                db.SaveChanges();
            }

            Logging.LogWriter.Write(String.Format("Updated Poll with PollID={0}", updatedPoll.PollID));
            return updatedPoll;
        }

        public void DeletePoll(int pollID)
        {
            Logging.LogWriter.Write(String.Format("Attempting to delete Poll with PollID={0}", pollID));

            using (var db = new Database())
            {
                Poll dbList = db.Polls.FirstOrDefault(x => x.PollID == pollID);
                if (dbList == null) throw new ArgumentException(String.Format("No Poll with PollID={0} exists", pollID));

                dbList.DeletedTime = DateTime.Now;
                db.SaveChanges();
            }

            Logging.LogWriter.Write(String.Format("Deleted Poll with ID {0}", pollID));
        }
    }
}
